formats/json/commonMain/src/kotlinx/serialization/json/internal/lexer/CommentLexers.kt

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2017-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.serialization.json.internal
+
+/*
+ Implementations of these two classes are nearly identical. However, there are several reasons why it can't be unified
+ and merged into one lexer:
+
+ 1. Making skipWhitespaces() a separate method slows deserialization down for about 10-15%. It is faster to have
+ `if (c == ' ' || c == '\n' || c == '\r' || c == '\t')` handling copy-pasted instead of separate method; however, copy-pasting comment
```

Comment:
Can you elaborate a bit more? Do you suggest a structure like this:
```
class StringLexerWithComments {
  fun consumeNextToken() {
    skipWhitespaces()
    ...
  }
  override fun skipWhitespaces() = ...
}

class StringLexer {
  fun consumeNextToken() {
    inlineSkipWhitespaces()
    ...
  }
  override fun skipWhitespaces() = inlineSkipWhitespaces()
  private inline fun inlineSkipWhitespaces() = ...
}
```

or something else?

Reply:
`consumeNextToken`s in `StringLexer` and `StringLexerWithComments` are identical modulo different `skipWhitespace` implementations. It seems to be true for few other methods as well.

So yeah, we can implement it like you mentioned:
```
class StringLexer {
  // this one will be the same for WithComments-variant, so we can write it only once.
  protected inline fun consumeNextTokenImpl(skipWs: StringLexer.() -> Int): Byte {
        val source = source
        val cpos = skipWs()
        if (cpos >= source.length || cpos == -1) return TC_EOF
        currentPosition = cpos + 1
        return charToTokenClass(source[cpos])
  }
  private inline skipWhitespaceImpl(): Int  { ... }
  
  override fun consumeNextToken() = consumeNextTokenImpl { skipWhitespaceImpl() }
  override fun skipWhitespaces() = skipWhitespaceImpl()
}

class StringLexerWithComment {
    private inline skipWhitespaceImpl(): Int  { ... }
    
    override fun consumeNextToken() = consumeNextTokenImpl { skipWhitespaceImpl() }
    override fun skipWhitespaces() = skipWhitespaceImpl()
}
```

and then we can use the same trick to share most of the skipWhitespace's logic between both WithComments classes.

But I changed my face to " 🤢 " somewhere in the middle of this message.

Reply:
I have a feeling that it wouldn't be very beneficial, as all `consume/peek/...` methods are slightly different, and replacing two of each with two one-liners and one implementation doesn't sound much

Reply:
@shanshin WDYT?

Reply:
> I have a feeling that it wouldn't be very beneficial, as all consume/peek/... methods are slightly different, and replacing two of each with two one-liners and one implementation doesn't sound much

Agree

Reply:
It doesn't seem to me that this will lead to an increase in readability,although at first glance it looks like in almost all functions it is possible to make a common code by skipping space characters